lib/fixed-points.nix

```
@@ -110,4 +110,94 @@ rec {
     fix' rattrs // {
       ${extenderName} = f: makeExtensibleWithCustomName extenderName (extends f rattrs);
    };
+
+  /*
+    Creates an overridable attrset with encapsulation.
+
+    This is like `makeExtensible`, but only the `public` attribute of the fixed
+    point is returned.
+
+    Synopsis:
+
+        r = encapsulate (final@{extend, ...}: {
+
+          # ... private attributes for `final` ...
+
+          public = {
+            # ... returned attributes for r, in terms of `final` ...
+            inherit extend; # optional, don't invoke too often; see below
+          };
+        })
+
+        s = r.extend (final: previous: {
```

```suggestion
        s = r.extend (this: old: {
```

I like this as a convention because it has a mnemonic: the reverse, "old this" can only be a description of one of the parameters and is therefore the wrong order.

We should probably not reuse names of overlays, so that overlays and `encapsulate` can be used together more easily. (For instance using an `haskellPackages` overlay inside a Nixpkgs overlay is quite common, and awkward.)
- `self: super:`: overlays
- `final: prev:`: Nixpkgs overlays
- `finalAttrs: prevAttrs:`: overrideAttrs
- `this: old:`: Things with overlay-like overriding and encapsulation.